docs/rfc/007-endpoint-errors.md

```diff
@@ -0,0 +1,229 @@
+should handle this outcome to ensure that you can react accordingly”. It should be something closely tied to the 
+functionality of that endpoint - that is to say, `getFoo` declaring `FooNotReadyYet` would be a good example, but 
+`UserNotAuthorized` would be undeclared / passed through as normal (as it is a standard expectation of interacting with 
+an authorized service).
+
+**Pros:**
+
+* Consumers now know the *expected* outcomes of calling an endpoint in a first-class way and can handle those outcomes accordingly.
+* Producers are able to express the different failure states of their endpoints in greater detail, and have conviction that those failure states are being handled by consumers.
+* Changing the failure states of an endpoint now visibly affects the API definition - changes in behavior are properly captured in code (rather than just documentation).
+
+**Cons:**
+
+* To be effective, these failure cases will need to be ‘forced’ upon consumers via the compiler. This could be as complex 
+  data types (perhaps unions of {success, error1, error2, ...} or via checked exceptions. This would entail some degree 
+  of (justified) work migrating to a with-errors world to wire each failure case into the correct behavior in client code.
+* Given the new flexibility, producers may more eagerly throw errors in their code, and ‘old’ clients will have a hard 
+  time dealing with these without upgrading.
+* Producers might be incentivized to cram *all* the possible outcomes of their endpoint as errors (see the 
+  `UserNotAuthorized` example above), and as such unduly burden consumers of their endpoint with things that should 
+  probably just be `RuntimeException`s. This in turn might have knock-on effects in the codebase depending on the exact
+  implementation used.
+* Introducing further types (whether as checked exceptions or otherwise) that determine control flow may result in them 
+  spreading throughout a code base and result in further coupling to the RPC layer. As an example, using checked exceptions
+  could encourage the consumer to throw the exception upwards rather than dealing with it at the callsite.
+  
+### Mandatory vs. Optional
+
+As discussed above, declared errors can be either made to be optional, or mandatory. Both versions come with their own
+pros and cons, and picking one side or the other inevitably involves making sacrifices.
+
+**The case for mandatory handling**
+
+_Forcing_ consumers to handle your exceptional cases is beneficial to the _producer_. You can have confidence that
+consumers will be encouraged to handle the exceptional cases, even if they are not reading your documentation. Acknowledging
+the presence of exceptional cases will be required at the minimum.
+
+There are benefits as a consumer as well - additional errors, behaviors, or edge cases will be exposed to you on upgrading
+your client, and you should (in theory) never miss out on a new case (assuming you keep your client relatively up to date).
```

Comment:
The way I've mentally modeled endpoint-errors is as a way to recommend that clients handle particular types of errors, not necessarily to enumerate all possible types of failures. How would you use the data from `endpoint xyz threw an undeclared internal precondition error`?
In some cases we don't expect a client to handle a failure, just to fail. I don't think we would want to force callers to enumerate cases that are inconsequential or equivalent. UNAUTHORIZED and FORBIDDEN ErrorCodes are remarkably common, and can be expected from any authed endpoint.

Reply:
> assuming you keep your client relatively up to date
do we have any data on how this looks in reality? I don't think we generally upgrade -api  dependencies, do we? (this may have changed)

Reply:
We do! Between the minimum-version calculation and pdeps-bot checks we've been able to start excavating api upgrades that trail releases by ~1 month by default.

Reply:
Internally we have an excavator that tried to keep clients 45 days behind the latest release

Reply:
so does this turn into also then people declare an error in their api, but don't feel ok with actually throwing it until `45 days + median service deploy time` have passed?

Reply:
Not quite. By leveraging the existing conjure error contract, existing clients will handle errors from new servers exactly the same way they do today. The wire format is identical, and all clients support parsing arbitrary conjure errors. Clients using sufficiently new generated code are able to advertise certain types of errors so that developers are inclined to handle them reasonably, and (better yet) avoid causing them in the first place.

Reply:
If you're adding state that must be parsed client-side to be correct there's only one option: make a new endpoint and add it to the return value. Outside of those, any solution will have its own set of downsides, where we can decide where those downsides land.

This feels like a reasonable trade-off, where we have a mechanism of rolling it out/making sure dev teams can realistically handle known errors, which should also make things like propagating human understandable error causes significantly easier. It does rely on some judgement on the API side on what constitutes an acceptable break and what doesn't.

No solution will be able to make clients forward compatible, nor should it try.

Reply:
On undeclared errors, the "how would you use the data" for me would be to be able to track which endpoints throw which errors. For errors thrown e.g. closer to the DB layer, I could accidentally introduce an error that I wanted to declare, but forgot to add to all endpoints.

Reply:
For this, I think we're likely just not using the right vocabulary to describe this problem. If an endpoint defines a new endpoint that may be thrown, there are probably a few cases that this can happen. I've probably missed a few.

For 5xx returns:
1. I'm switching from a generic server error to a specific server error. Generally, expect you shouldn't be handling 5xx returns.

For 4xx returns
1. I'm switching from a generic client error to a specific client error. This is clearly not API in this model, clients needn't take an action because they shouldn't have been depending on that behaviour.
2. I used to throw one client error, and am changing to throw a different one (e.g. NotFound to DatasetNotFound).
3. I added a new argument to my method, and if you use it improperly you get a MyNewError. Old consumers probably needn't care about this.

The only cases we probably should care about for the evolution component is 2. This is clearly an API break, so conjure backcompat should not let you add that extra declared argument. Stuff around pdeps bot is pretty immaterial - we have an API contract that says that old code should not stop working, we need to maintain that. The encouragement that we can make changes after 45 days is only reasonable once we bump the major version every 45 days - in absence we'll just introduce a new class of weird errors that will happen regularly that the system used to but no longer protects against.

One way around this is to let the server throw an error which explicitly has a 'parent' error, letting one throw the new error but indicate that it used to be an old error. Therefore if the client does not know about the new error, they can process according to the old code.